### ISSUE TITLE
sql: more precise resulting type calculation for built-ins

### ISSUE THREAD

**Issue by Korablev77** [tarantool/tarantool#4456]
Currently, such functions as `greatest`, `least`, `min`, `max`, `nullif` and `coalesce` are supposed to feature `'SCALAR'` return type. It affects type which is set in query's meta-information:

```
SELECT GREATEST('a', 'b', 'c');
---
- metadata:
  - name: GREATEST('a', 'b', 'c')
    type: scalar
  rows:
  - ['c']
...
```
Type in metadata is calculated during query preparation, so it can't be always determined precisely. For instance, without query execution it is impossible to point out resulting type of `CASE-THEN` statement if at least two `THEN` clauses return values of different types. On the other hand, if all arguments are of the same type, we can return that type instead of more general `'SCALAR'`

**Comment by pgulutzan** [comment]
The query meta-information is also incorrect for SELECT NULL;.

**Comment by ImeevMA** [comment]
`COALESCE`, `NULLIF` and `IFNULL` will be reworked in issue #6363 . This issue is closed by 0f144350af95780b5981b657e1bc590d48206e0a.
